### PR TITLE
Prevent creating checkout with unpublished product

### PR DESCRIPTION
**Pull request by d-wysocki**
I want to merge this change because...closes https://github.com/mirumee/saleor/issues/5676

# Impact

* [ ] New migrations
* [ ] New/Updated API fields or mutations
* [ ] Deprecated API fields or mutations
* [ ] Removed API types, fields, or mutations

# Pull Request Checklist

* [X] Privileged queries and mutations are guarded by proper permission checks
* [X] Database queries are optimized and the number of queries is constant
* [X] Database migration files are up to date
* [X] The changes are tested
* [X] GraphQL schema and type definitions are up to date
* [ ] Changes are mentioned in the changelog

### CODE REVIEW

File: saleor/graphql/checkout/mutations.py [hunk 1]
```diff
@@ -283,7 +283,11 @@ def save(cls, info, instance: models.Checkout, cleaned_input):
                     raise ValidationError(
                         f"Insufficient product stock: {exc.item}", code=exc.code
                     )
-
+                except ProductNotPublished as exc:
+                    raise ValidationError(
+                        "Can't create checkout with unpublished product.",
```
Comment (korycins) [inline saleor/graphql/checkout/mutations.py:288]: AFAIK Checkout can be already created. This should be something like `Can't add ...`

File: tests/api/test_checkout.py [hunk 2]
```diff
@@ -754,6 +754,28 @@ def test_checkout_lines_add(
...
+    variables = {
+        "checkoutId": checkout_id,
+        "lines": [{"variantId": variant_id, "quantity": 1}],
+    }
+
+    response = user_api_client.post_graphql(MUTATION_CHECKOUT_LINES_ADD, variables)
+
+    content = get_graphql_content(response)
+    error = content["data"]["checkoutLinesAdd"]["errors"][0]
+    assert error["message"] == "Can't create checkout with unpublished product."
```
Comment (korycins) [inline tests/api/test_checkout.py:776]: Can you add assert for `error['code']` here?

**Comment by fowczarek** [comment]
How about `checkoutLinesUpdate` mutation with unpublished products?

@maarcingebala  Should we allow to add unpublished products to draft order? 
If yes, should we allow to finalize the draft order with unpublished products?

File: tests/api/benchmark/test_checkout_mutations.py [hunk 3]
```diff
@@ -228,6 +228,50 @@ def test_create_checkout(
     assert checkout_counts + 1 == Checkout.objects.count()
 
 
+@pytest.mark.django_db
+@pytest.mark.count_queries(autouse=False)
+def test_create_checkout_with_unpublished_product(
```
Comment (fowczarek) [inline tests/api/benchmark/test_checkout_mutations.py:233]: Why this test is in `benchmark` folder?

Comment (d-wysocki) [reply tests/api/benchmark/test_checkout_mutations.py:233]: In this folder there are also other tests related to checkout like `test_create_checkout`, `test_update_checkout_lines.

**Comment by maarcingebala** [comment]
@fowczarek I think it shouldn't be possible to add unpublished products to draft orders as well. At least for now it's safer to assume this approach, and if we have a case that we need to add them to draft orders, then we'll make sure that it's properly implemented.

File: saleor/graphql/checkout/mutations.py [hunk 4]
```diff
@@ -354,6 +358,11 @@ def perform_mutation(cls, _root, info, checkout_id, lines, replace=False):
                     raise ValidationError(
                         f"Insufficient product stock: {exc.item}", code=exc.code
                     )
+                except ProductNotPublished as exc:
+                    raise ValidationError(
+                        "Can't create checkout with unpublished product.",
```
Comment (fowczarek) [inline saleor/graphql/checkout/mutations.py:363]: This isn't `checkout_create` mutation. :police_officer:

**Pull request merged**